### PR TITLE
Implement Directions utility class

### PR DESCRIPTION
**Pull request by tttzach**
This class was created prior to client abstraction which will be addressed on #46 
This pull request replaces #25 to remove all unrelated commit history

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  public static void sendJson(HttpServletResponse response, Object object) throws IOException {
+    Gson gson = new Gson();
+    String json = gson.toJson(object);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  /** Get directions using the parameters stored in this class. */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (AymarMuhikira) [inline src/main/java/com/google/sps/servlets/DirectionsServlet.java:49]: Nit: you could add the parameters of the method in the Javadoc description

File: src/main/java/com/google/sps/utility/DirectionsUtility.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   * @param destination A string representing the destination to get directions to.
+   * @param origin A string representing the origin to get directions from.
+   * @param waypoints A list of string consisting of waypoints to visit between the destination and
+   *     the origin.
+   * @param apiKey A string representing the API key to authenticate a Google Directions API call.
+   * @return A string representing the URI.
+   * @throws URISyntaxException Indicates that the built string could not be parsed as a URI.
+   */
+  public static String getDirectionsUri(
+      String destination, String origin, List<String> waypoints, String apiKey)
```
Comment (AymarMuhikira) [inline src/main/java/com/google/sps/utility/DirectionsUtility.java:61]: Does this method need to be accessible outside of the class?

File: src/main/webapp/directions.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* exported getDirections */
+
+/**
+ * Gets key information from each leg and displays them on the DOM.
+ */
+function getDirections() {
```
Comment (AymarMuhikira) [inline src/main/webapp/directions.js:20]: Consider adding this to the populateData script.

File: src/test/java/DirectionsUtilityTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+@RunWith(JUnit4.class)
+public class DirectionsUtilityTest {
+
+  private final String ORIGIN = "University of Waterloo";
+  private final String DESTINATION = "Wilfrid Laurier University";
+  private final List<String> NO_WAYPOINTS = Arrays.asList();
+  private final List<String> ONE_WAYPOINT = Arrays.asList("University of Waterloo Place");
+  private final List<String> MANY_WAYPOINTS =
+      Arrays.asList("University of Waterloo Place", "Google Kitchener");
+  private final String API_KEY = "SAMPLE_API_KEY";
```
Comment (AymarMuhikira) [inline src/test/java/DirectionsUtilityTest.java:34]: Some other edge cases to consider: what if one of the end points is also a waypoint, what if  a waypoint is duplicated in the many waypoints scenario, is it possible to have an unrecognized address? What are the expected behaviours for these cases? Should we test for those>

Comment (tttzach) [reply src/test/java/DirectionsUtilityTest.java:34]: These are all valid edge cases! All of these should be handled in the step before using this utility which involves the Google Places API (handles all the unrecognized addresses) and creates only valid origin, destination and waypoints parameters. I should definitely note this on the file, thanks!

File: src/main/webapp/directions.js (hunk 3 above)
Comment (tttzach) [reply src/main/webapp/directions.js:20]: This is a great idea, thanks!

File: src/main/java/com/google/sps/utility/DirectionsUtility.java (hunk 2 above)
Comment (tttzach) [reply src/main/java/com/google/sps/utility/DirectionsUtility.java:61]: Nope, the only reason this is public was only for testing purposes. Should probably test these within this class. Thanks for pointing it out!

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java (hunk 1 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/DirectionsServlet.java:49]: Absolutely, thanks for catching that!

File: src/main/webapp/directions.js (hunk 3 above)
Comment (tttzach) [reply src/main/webapp/directions.js]: `RESOLVED`

File: src/test/java/DirectionsUtilityTest.java (hunk 4 above)
Comment (tttzach) [reply src/test/java/DirectionsUtilityTest.java:38]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+@WebServlet("/directions")
+public class DirectionsServlet extends HttpServlet {
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/servlets/DirectionsServlet.java:31]: Public classes require javadoc. [7.3](https://google.github.io/styleguide/javaguide.html#s7.1-javadoc-formatting)

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    if (waypointsString == null) {
+      waypoints = Arrays.asList();
+    } else {
+      waypoints = Arrays.asList(waypointsString.split(";"));
+    }
+    try {
+      JSONObject directions =
+          DirectionsUtility.getDirections(destination, origin, waypoints, apiKey);
+      sendJson(response, directions);
+    } catch (JSONException e) {
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/servlets/DirectionsServlet.java:70]: These two catch blocks can be simplified by doing: 

`catch (JSONException | URISyntaxException e) {...}`

File: src/main/java/com/google/sps/utility/DirectionsUtility.java (hunk 2 above)
Comment (austinteshuba) [reply src/main/java/com/google/sps/utility/DirectionsUtility.java:61]: If there could be a use for this outside of the class, you could consider making this public.

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java (hunk 6 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/DirectionsServlet.java:70]: Thanks for letting me know, I have incorporated this in the new dependency injection ready PR #63 . Thanks!

File: src/main/java/com/google/sps/servlets/DirectionsServlet.java (hunk 5 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/DirectionsServlet.java:31]: I have incorporated this in the new dependency injection ready PR #63 . Thanks!

**Comment by tttzach** [comment]
This PR is now a duplicate of #63 since #63 reimplements this pipeline to the Google Directions API in a structure which allows for dependency injection.
